crates/orderbook/src/solvable_orders.rs

```diff
@@ -327,41 +334,64 @@ impl Maintaining for SolvableOrdersCache {
 async fn get_orders_with_native_prices(
     mut orders: Vec<Order>,
     native_price_estimator: &dyn NativePriceEstimating,
+    deadline: Instant,
 ) -> (Vec<Order>, BTreeMap<H160, U256>) {
     let traded_tokens = orders
         .iter()
         .flat_map(|order| [order.creation.sell_token, order.creation.buy_token])
         .collect::<HashSet<_>>()
         .into_iter()
         .collect::<Vec<_>>();
-    let prices = native_vec_estimates(native_price_estimator, &traded_tokens)
-        .await
-        .into_iter()
-        .zip(traded_tokens)
-        .filter_map(|(price, token)| match price {
-            Ok(price) => Some((token, to_normalized_price(price)?)),
-            Err(err) => {
-                tracing::warn!(?token, ?err, "error estimating native token price");
-                None
-            }
-        })
-        .collect::<BTreeMap<_, _>>();
+    let mut prices = BTreeMap::new();
+    let mut price_stream = native_price_estimator.estimate_native_prices(&traded_tokens);
+    let collect_prices = async {
+        while let Some((index, result)) = price_stream.next().await {
+            let token = &traded_tokens[index];
+            let price = match result {
+                Ok(price) => price,
+                Err(err) => {
+                    tracing::warn!(?token, ?err, "error estimating native token price");
+                    continue;
+                }
+            };
+            let price = match to_normalized_price(price) {
+                Some(price) => price,
+                None => continue,
+            };
+            prices.insert(*token, price);
+        }
+    };
+    match tokio::time::timeout_at(deadline, collect_prices).await {
+        Ok(()) => (),
+        Err(_) => tracing::warn!(
+            "auction native price collection took too long, got {} out of {}",
+            prices.len(),
+            traded_tokens.len()
+        ),
+    }
 
+    // Filter both orders and prices so that we only return orders that have prices and prices that
+    // have orders.
```

Comment:
Ah! This is important because we may have gotten the price for one of the tokens of an order, but not the other, and the token we have the price for only appears in that order. I.e. we are missing a token for an order, so we exclude the order; in excluding the order, we have a price with no equivalent order.